source/common/stream_info/set_filter_state_object_impl.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include "envoy/stream_info/set_filter_state_object.h"
+
+#include "absl/container/flat_hash_set.h"
+
+namespace Envoy {
+namespace StreamInfo {
+
+/**
+ * Implementation of SetFilterStateObject.
+ */
+template <typename T> class SetFilterStateObjectImpl : public SetFilterStateObject<T> {
+public:
+  void add(const T& address) override { values_.emplace(address); }
+
+  void clear() override { values_.clear(); }
+
+  void iterate(const std::function<bool(const T& address)>& fn) const override {
+    for (const auto& address : values_) {
+      if (!fn(address)) {
+        break;
+      }
+    }
+  }
+
+  static const std::string& key() {
+    CONSTRUCT_ON_FIRST_USE(std::string, "filter_state_key.set_object");
```

Comment:
This name doesn't make sense from a documentation perspective. There should be a name specific to the call site where it's used (something about upstream address data, etc.)

Reply:
> This name doesn't make sense from a documentation perspective. There should be a name specific to the call site where it's used (something about upstream address data, etc.)

Since this is a generic structure which is a set of `T`,  should it have reference to the use case? Also, do you think the  following container definition is ok:

``` 
template <typename T> class SetFilterStateObjectImpl : FilterStateObject {
public:
// accessor and mutator
private:
  absl::flat_hash_set<T> values_;
};

```

This should be generic enough to be used elsewhere.

Reply:
My point is the filter state is part of the documentation, so it needs to be relevant. So what I would do is just make a derived class that defines the name only, and then just have a public `absl::flat_hash_set<T> values_`. It doesn't even need to be private.

Reply:
> My point is the filter state is part of the documentation, so it needs to be relevant. So what I would do is just make a derived class that defines the name only, and then just have a public `absl::flat_hash_set<T> values_`. It doesn't even need to be private.

Thanks @mattklein123 . 
If we have a name specific to `upstream address`, I am worried we will lose the generic container usage. One way to get around it could be to remove the `key` from the class and have it defined in a common place. This key can then be referenced by `rbac` and `dynamic forward proxy`. I want to be also careful about  `dynamic forward proxy ` not directly referring anything in `rbac`

Reply:
I think you are over thinking this. :) I would purely just have an overridden name or just define a struct specific to this use case.

Reply:
Yeah, see my API comment, can't we just have an `RbacUpstreamIp` filter state, `RbacUpstreamPort` etc? This would be much simpler.

The idea of filter state is this is what you are posting to; so, dynamic forward is able to write to this state, which is then read by some RBAC filter.